Add Cmd+Shift+F as tidy code shortcut and prevent browser default behaviour

vulongphan opened the pull request:
Fixes #1711

I have verified that this pull request:

* [x] has no linting errors (`npm run lint`)
* [x] is from a uniquely-named feature branch and has been rebased on top of the latest `develop` branch. (If I was asked to make more changes, I have made sure to rebase onto `develop` then too)
* [x] is descriptively named and links to an issue number, i.e. `Fixes #123`

vulongphan committed:
Add Cmd+Shift+F astidy code shortcut and prevent browser default behaviour

vulongphan commented:
@catarak what I did was to detect key short cut of <kbd>Cmd</kbd>+<kbd>Shift</kbd>+<kbd>F</kbd> to call `tidyCode()` and to prevent browser default behaviour (Full Screen Mode in Chrome on Mac). Could you please have a look and tell me if I have to change anything?

catarak commented on client/modules/IDE/components/Editor.jsx:
```diff
@@ -155,14 +160,21 @@ class Editor extends React.Component {
     this._cm.on('keydown', (_cm, e) => {
-      // 9 === Tab
-      if (e.keyCode === 9 && e.shiftKey) {
+      this.handleKey(this.map, e);
```
I'm not sure how using this function `this.handleKey` is helping in this scenario? Just from reading this it seems like it's adding some intermediate code that is unnecessary, but please explain if I'm wrong!

vulongphan replied:
The reason why I use this function is because to detect shortcut <kbd>Cmd</kbd>+<kbd>Shift</kbd>+<kbd>F</kbd> the following conditional statement would not work
```
if (e.keyCode === 91 && e.keyCode === 16 && e.keyCode === 70){ }
```
I guess the reason is because the browser cannot detect multiple key press. To solve this, I used an object called `this.map` to store the `keydown` state of the keyboard. For example, if the shortcut <kbd>Cmd</kbd>+<kbd>Shift</kbd>+<kbd>F</kbd> is pressed, `this.map` will be
```
{91: true, 16: true, 70: true}
```
 and when we release the shortcut, `this.map` will be
```
{91: false, 16: false, 70: false}
```
I wonder if there is a better way to get around this.

catarak replied:
there is! if you look at the documentation for a [KeyboardEvent](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent), you'll see that it contains `metaKey`, which will be true if <kbd>Command</kbd> is pressed, `ctrlKey`, which will be true if <kbd>Control</kbd> is pressed, and `shiftKey`, which will be true if <kbd>Shift</kbd> is pressed.

catarak commented on client/modules/IDE/components/Editor.jsx:
```diff
@@ -155,14 +160,21 @@ class Editor extends React.Component {
-      // 9 === Tab
-      if (e.keyCode === 9 && e.shiftKey) {
+      this.handleKey(this.map, e);
+      // 91 === Cmd
```
There's a variable that's imported to this file called `metaKey` which will be set to `'Cmd'` if the user is on Mac, and `'Ctrl'` if the user is on Windows/Linux. I think you should use this variable rather than just `'Cmd'` since `'Cmd'` doesn't exist on Windows/Linux!

vulongphan replied:
Yes, I will incorporate the use of `metaKey` in my changes!

catarak commented on client/modules/IDE/components/Editor.jsx:
```diff
@@ -62,7 +62,11 @@ const beautifyHTML = beautifyJS.html;
+
+// delete CodeMirror.keyMap.sublime['Shift-Cmd-F'];
+
+console.log(CodeMirror.keyMap.sublime);
```
Could you delete this `console.log` and the commented out code?

vulongphan replied:
Yes!

vulongphan committed:
Merge remote-tracking branch 'upstream/develop' into update/tidy-code-shortcut

vulongphan committed:
make use of variable metaKey

vulongphan commented:
I have made the changes accordingly, could you please have a look?

Thank you!

catarak committed:
Merge develop

vulongphan commented:
@catarak I am sorry but does this mean that this PR should be closed now? Please correct me if I missed something here

catarak commented:
@vulongphan I just added a comment about using `KeyboardEvent.metaKey`, `KeyboardEvent.shiftKey`, and `KeyboardEvent.ctrlKey`—could you use these rather than `handleKey`?

vulongphan committed:
use KeyboardEvent to check for shortcut to tidyCode

vulongphan committed:
use KeyboardEvent to check for shortcut to tidyCode

vulongphan committed:
use KeyboardEvent to check for shortcut to tidyCode

vulongphan commented:
@catarak It was my bad I did not see your comment earlier on. Also please ignore the first two commits ([e53f956](https://github.com/processing/p5.js-web-editor/pull/1732/commits/e53f9561ec43d76d2b65a0c1c0cd77cc030c0142) and [0be0c85](https://github.com/processing/p5.js-web-editor/pull/1732/commits/0be0c854501869bfa1660fc66ba1ed4f1c35dd41)) and check [059b171](https://github.com/processing/p5.js-web-editor/pull/1732/commits/059b1710c74058bf5801df4d778cc9215bd5a87c).
I am really sorry for the inconvenience.

catarak commented:
> @catarak It was my bad I did not see your comment earlier on. Also please ignore the first two commits ([e53f956](https://github.com/processing/p5.js-web-editor/pull/1732/commits/e53f9561ec43d76d2b65a0c1c0cd77cc030c0142) and [0be0c85](https://github.com/processing/p5.js-web-editor/pull/1732/commits/0be0c854501869bfa1660fc66ba1ed4f1c35dd41)) and check [059b171](https://github.com/processing/p5.js-web-editor/pull/1732/commits/059b1710c74058bf5801df4d778cc9215bd5a87c).
> I am really sorry for the inconvenience.

No need to apologize!!! Contributing is about learning rather than being perfect 😄

catarak commented on client/modules/IDE/components/Editor.jsx:
```diff
@@ -168,8 +167,9 @@ class Editor extends React.Component {
-      // 9 === Tab
-      if (e.keyCode === 9 && e.shiftKey) {
+      // 70 === f
+      if ((e.metaKey || e.ctrlKey) && e.shiftKey && e.keyCode === 70) {
```
i think it would also make sense to check the value of [`metaKey`](https://github.com/processing/p5.js-web-editor/blob/059b1710c74058bf5801df4d778cc9215bd5a87c/client/modules/IDE/components/Editor.jsx#L39), so that Ctrl + Shift + F doesn't work on Mac.

vulongphan replied:
Yes, I have added the check accordingly!

vulongphan committed:
use metaKey to check for Windows/MacOs

catarak commented:
Working great for me! Would you mind updating the [KeyboardShortcutModal](https://github.com/processing/p5.js-web-editor/blob/develop/client/modules/IDE/components/KeyboardShortcutModal.jsx) component with the new shortcut? Thank you!

vulongphan committed:
update TidyCode shortcut modal

vulongphan commented:
Hi @catarak, I have updated the file but for some reason I cannot see the change in my local development. Could you please check for me? Attached is the local development after I updated the file
![Screen Shot 2021-02-18 at 5 13 10 PM](https://user-images.githubusercontent.com/46087559/108361970-9599f100-720c-11eb-83b9-9d20f348f108.png)

catarak committed:
Merge branch 'develop' into update/tidy-code-shortcut

catarak commented:
I'm seeing the same thing! I think the issue is that you need to update the file [Nav.jsx](https://github.com/processing/p5.js-web-editor/blob/develop/client/components/Nav.jsx#L425) as well.

vulongphan committed:
Merge remote-tracking branch 'upstream/develop' into update/tidy-code-shortcut
Merge upstream/develop

vulongphan committed:
update Nav bar with the short cut for tidy code

vulongphan committed:
Merge branch 'update/tidy-code-shortcut' of https://github.com/vulongphan/p5.js-web-editor into update/tidy-code-shortcut
Pulled from origin update/tidy-code-shortcut

vulongphan commented:
@catarak yes I really do agree that we have to update `Nav.jsx` as well. I have updated the file and now see the changes in my local development.
However, `npm run test` has failed which does not seem to be caused by the changes I have made. Could you please give me some suggestion?

Thank you!

catarak commented:
Because you're changing the appearance of `Nav.jsx`, you need to update the snapshot as well: `npm run test -- --updateSnapshot`. I'm going to go ahead and just do this so I can merge this fix in.

catarak committed:
[#1711] Update Nav snapshot
